addons/survey/controllers/main.py

```diff
@@ -732,35 +727,130 @@ def _get_user_input_domain(self, survey, line_filter_domain, **post):
+
+    def _get_matching_user_input_lines(self, filters_domains, user_input_domain):
+        """ Retrieve the matching user input lines depending on the survey base domain and the filters domains.
+
+        :return recordset: all matching user input lines
+        """
+        # Add filters domain to the base domain
+        if filters_domains:
+            all_required_lines_domains = [
+                [('user_input_line_ids', 'in', request.env['survey.user_input.line'].sudo()._search(filters_domain))]
+                for filters_domain in filters_domains
+            ]
+            user_input_domain = expression.AND([user_input_domain, *all_required_lines_domains])
+
+        # Get the matching user input lines
+        user_inputs_query = request.env['survey.user_input'].sudo()._search(user_input_domain)
+        user_input_lines = request.env['survey.user_input.line'].search([('user_input_id', 'in', user_inputs_query)])
+
+        return user_input_lines
+
+    @staticmethod
+    def _get_filters_from_post(post):
+        """ Extract the filters from post depending on the model that needs to be called to retrieve the filtered answer data.
+
+        * Model = 'A' : Match a `survey.question.answer` record (simple_choice, multiple_choice, matrix)
+        * Model = 'L' : Match a `survey.user_input.line` record (char_box, text_box, numerical_box, date, datetime)
+        :rtype: (collections.defaultdict[int, list[int]], list[int])
+        """
+        answer_cols_to_rows_mapping = defaultdict(list)
+        answer_user_input_line_ids = []
+
+        for data in post.get('filters', '').split('|'):
+            if not data:
+                break
+            model, row_id, ans_id = data.split(',')
+            row_id, ans_id = int(row_id), int(ans_id)
+            if model == 'A':
+                answer_cols_to_rows_mapping[ans_id].append(row_id)
+            elif model == 'L' and not row_id:
+                answer_user_input_line_ids.append(ans_id)
+
+        return answer_cols_to_rows_mapping, answer_user_input_line_ids
+
+    @staticmethod
+    def _get_value_with_operator(record, row_id=0):
+        """ Get the answer's value and domain operator to find answers with the
+        same value (depends on answer's model and question type).
+
+        :param record: 'survey.question.answer' or 'survey.user_input.line' record
+        :rtype: (string, string)
+        """
+        if record.question_id.question_type in ('simple_choice', 'multiple_choice', 'matrix'):
+            return f'{f"{row_id.value} : " if row_id else ""}{record.value}', None
+        return record[f'value_{record.question_id.question_type}'], {
+            'text_box': 'ilike',
+            'char_box': '=ilike',
+            'date': '=',
+            'datetime': '=',
+            'numerical_box': '=',
+        }[record.question_id.question_type]
+
+    @staticmethod
+    def _prepare_matching_answer_domain(question_type, answer, row_id=0, operator=None, value=None):
```

Comment:
So this receives both survey.question.answer and survey.user_input.line records and it works because they share common fields?

-> :exploding_head: 

Generally we tend to move as much logic as possible into the models themselves.

So here I would duplicate this "_prepare_matching_answer_domain" in both models and make it work on "self" (with an ensure_one). That way we should have less parameter and more specific functions, even if some domain code is duplicated.